File: 2022/deno/04.ts
const input = await Deno.readTextFile("./input/04.in");

const data = input.split("\n");

/**
 * Part One
 */

const overlaps = (left: number[], right: number[]) => {
  return left[0] <= right[0] && right[1] <= left[1];
};

const groups = data.map((group) => {
  const [left, right] = group.split(",");

  const leftRange = left.split("-").map(Number);
  const rightRange = right.split("-").map(Number);

  return [leftRange, rightRange];
});

console.log(
  "Part one:",
  groups.filter(
    (group) => overlaps(group[0], group[1]) || overlaps(group[1], group[0])
  ).length
);

/**
 * Part Two
 */

const anyOverlap = (left: number[], right: number[]) => {
  return right[0] <= left[1] && right[1] >= left[0];
};

console.log(
  "Part two:",
  groups.filter(
    (group) => anyOverlap(group[0], group[1]) || anyOverlap(group[1], group[0])
  ).length
);

File: 2022/swift/04.swift
import Foundation

let filename = "./input/04.in"

func totalOverlap(_ left: [Int], _ right: [Int]) -> Bool {
    left[0] <= right[0] && right[1] <= left[1]
}

func partialOverlap(_ left: [Int], _ right: [Int]) -> Bool {
    right[0] <= left[1] && right[1] >= left[0]
}

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let blocks = contents.components(separatedBy: "\n").map { $0.split(separator: ",").map {
            block in
            block.split(separator: "-").map { Int($0)! }
        }}

        print("Part one:", blocks.filter { totalOverlap($0[0], $0[1]) || totalOverlap($0[1], $0[0]) }.count)
        print("Part two:", blocks.filter { partialOverlap($0[0], $0[1]) || partialOverlap($0[1], $0[0]) }.count)

    } catch {
        print(error)
    }
}

main()
